[firmware] Fix HTTP 500 when saving device after credentials are removed

komaleswarreddy opened the pull request:
This PR fixes an issue where deleting device credentials after assigning
a firmware image causes Django admin to crash with HTTP 500.

The crash was caused by an uncaught DoesNotExist exception raised during
firmware upgrade validation when no valid device connection was found.
This PR converts the exception into a ValidationError so the admin shows
a proper, recoverable error message.
 

 Changes
- Catch missing device connection during firmware upgrade validation
- Raise ValidationError instead of DoesNotExist
- Add regression tests covering UpgradeOperation and DeviceFirmware

Fixes #250


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved validation and clearer, user-facing error messages when firmware upgrade options or device credentials are missing or removed; upgrade-option errors are converted to string-safe validation responses.

* **Tests**
  * Added regression tests ensuring ValidationError (not missing-object errors) is raised with helpful messages when credentials are unavailable.

* **Chores**
  * Simplified test environment: switched test DB backend, disabled GIS/notification test components, and added a test dependency for notifications.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

komaleswar committed:
[firmware] Prevent 500 error when device credentials are missing during firmware validation #250

Convert DeviceConnection.DoesNotExist raised during firmware upgrade
validation into a proper ValidationError so the admin form can
display a recoverable error instead of crashing.

This allows users to reassign credentials and save the device
configuration without encountering an internal server error.

Fixes #250
.

pandafy commented on openwisp_notifications/migrations/0001_initial.py:
I still see these files.

nemesifier commented on openwisp_notifications/migrations/0001_initial.py:
```diff
@@ -0,0 +1,10 @@
+# Minimal stub migration for openwisp_notifications
```
Adding `openwisp_notifications` here is most likely wrong, what is prompting you to do this?

nemesifier reviewed (changes requested):
In the current state this PR adds more problems than it solves.

If the CI build is failing for some reason, let it fail so we can debug it instead of trying to cover it up with bloat.

See my comments below.

Please remove the extra bloat.

nemesifier commented:
@coderabbitai review

komaleswar committed:
[firmware_upgrader] Prevent admin crash when credentials are missing #250

Convert DeviceConnection.DoesNotExist raised during firmware upgrade
validation into ValidationError.

This prevents HTTP 500 errors in Django admin when device credentials
are removed and allows users to recover.

Also removes openwisp_notifications stub that was causing migration
dependency issues.

Fixes #250

nemesifier committed:
Merge branch 'master' into issues/250-handle-missing-credentials-validation

nemesifier commented on requirements-test.txt:
```diff
@@ -1,4 +1,5 @@
 openwisp-utils[qa,selenium] @ https://github.com/openwisp/openwisp-utils/archive/refs/heads/1.3.tar.gz
+openwisp-notifications @ https://github.com/openwisp/openwisp-notifications/archive/refs/heads/1.3.tar.gz
```
Thanks for the contribution! However, we cannot accept changes to dependencies or core settings at this time as they complicate our CI/CD pipeline. Please revert those specific changes so we can focus on reviewing the core logic. Since we’re currently handling a high volume of PRs, we have to prioritize "clean" submissions that don't introduce build risks. If this task feels a bit complex, I’d be happy to point you toward some "Good First Issues" that are easier to get through the pipeline!
